const ErrorHandler = require("../utils/ErrorHandler");
const catchAsyncError = require("./catchAsyncError");
const jwt = require("jsonwebtoken");
const User = require("../model/user");  
const Shop = require('../model/shop')

exports.isAuthenticated = catchAsyncError(async (req, resp, next) => {
  const { token } = req.cookies;

  if (!token) {
    return next(new ErrorHandler("Please login to continue", 401));  
  } 

  const decoded = jwt.verify(token, process.env.JWT_SECRET_KEY);

  const user = await User.findById(decoded.id);

  if (!user) {
    return next(new ErrorHandler("User not found", 404)); 
  }

  req.user = user;  
  next();
});


exports.isSeller= catchAsyncError(async (req, resp, next) => {
  const { seller_token } = req.cookies;

  if (!seller_token) {
    return next(new ErrorHandler("Please login to continue", 401));  
  } 

  const decoded = jwt.verify(seller_token, process.env.JWT_SECRET_KEY);

  const shop = await Shop.findById(decoded.id);

  if (!shop) {
    return next(new ErrorHandler("Shop not found", 404)); 
  }

  req.seller = shop;  

  next();
});



exports.isAdmin = (...roles) => {
  return (req,res,next) => {
    if(!roles.includes(req.user.role)){
      return next (new ErrorHandler(`${req.user.role} can not access this resources!`))
    }
    next()
  }
}